File: python/src/frtb/sa.py
"""Standardised Approach assembly: SBM charges per risk class and scenario,
plus DRC-lite and RRAO.  SA capital = SBM + DRC + RRAO.
"""
from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Dict, List, Mapping, Optional, Sequence, Tuple

from .instruments import Bond, Instrument
from .market import Market
from .params import SbmParams
from .pricers import price_bond
from .sbm import SCENARIOS, RiskClassCharge, curvature_charge, delta_vega_charge
from .sensitivities import Sensitivities

RISK_CLASSES = ("girr", "equity", "fx")
MEASURES = ("delta", "vega", "curvature")


@dataclass(frozen=True)
class SbmResult:
    """SBM capital with full drill-down.

    charges[risk_class][measure][scenario] -> charge
    kb_medium[risk_class][measure] -> {bucket: K_b} (medium scenario)
    scenario_totals[scenario] -> sum over risk classes and measures
    capital = max over scenarios of scenario_totals.
    """

    charges: Dict[str, Dict[str, Dict[str, float]]]
    kb_medium: Dict[str, Dict[str, Dict[str, float]]]
    scenario_totals: Dict[str, float]
    capital: float


def _girr_structs(sens: Sensitivities, params: SbmParams):
    """GIRR delta WS per currency bucket + tenor-index correlation lookup."""
    tenor_index = {t: i for i, t in enumerate(params.girr_tenors)}
    bucket_ws = {
        ccy: {f"{t:g}": params.girr_rw(t) * s for t, s in per_tenor.items()}
        for ccy, per_tenor in sens.girr.items()
    }

    def intra_rho(_bucket: str, k: str, l: str) -> float:
        return params.girr_rho_kl(tenor_index[float(k)], tenor_index[float(l)])

    def gamma(_b: str, _c: str) -> float:
        return params.girr_gamma

    return bucket_ws, intra_rho, gamma


def _equity_structs(sens_map: Mapping[str, float], market: Market, params: SbmParams,
                    vega: bool):
    """Equity delta or vega WS per bucket, factor key = underlier name."""
    bucket_ws: Dict[str, Dict[str, float]] = {}
    for name, s in sens_map.items():
        b = market.equity(name).bucket
        p = params.equity_bucket(b)  # ValueError if the bucket is not pinned
        rw = p.vega_rw if vega else p.delta_rw
        bucket_ws.setdefault(b, {})[name] = rw * s

    def intra_rho(bucket: str, _k: str, _l: str) -> float:
        return params.equity_bucket(bucket).rho

    def gamma(_b: str, _c: str) -> float:
        return params.equity_gamma

    return bucket_ws, intra_rho, gamma


def _fx_structs(sens_map: Mapping[str, float], params: SbmParams):
    """FX WS: single pinned bucket 'FX', factor key = currency pair."""
    bucket_ws = {"FX": {pair: params.fx_delta_rw * s for pair, s in sens_map.items()}} \
        if sens_map else {}

    def intra_rho(_b: str, _k: str, _l: str) -> float:
        return params.fx_rho

    def gamma(_b: str, _c: str) -> float:
        return params.fx_gamma

    return bucket_ws, intra_rho, gamma


def sbm_capital(sens: Sensitivities, market: Market, params: SbmParams) -> SbmResult:
    """Assemble the full SBM capital: 3 risk classes x 3 measures x 3 scenarios."""
    hi, lo = params.scenario_high, params.scenario_low
    charges: Dict[str, Dict[str, Dict[str, float]]] = {rc: {m: {} for m in MEASURES}
                                                       for rc in RISK_CLASSES}
    kb_medium: Dict[str, Dict[str, Dict[str, float]]] = {rc: {} for rc in RISK_CLASSES}

    # -- delta / vega -------------------------------------------------------
    girr_ws, girr_rho, girr_gamma = _girr_structs(sens, params)
    eqd_ws, eq_rho, eq_gamma = _equity_structs(sens.equity_delta, market, params, vega=False)
    eqv_ws, _, _ = _equity_structs(sens.equity_vega, market, params, vega=True)
    fx_ws, fx_rho, fx_gamma = _fx_structs(sens.fx_delta, params)

    dv_specs = [
        ("girr", "delta", girr_ws, girr_rho, girr_gamma),
        ("girr", "vega", {}, girr_rho, girr_gamma),  # no IR-vol instruments in scope
        ("equity", "delta", eqd_ws, eq_rho, eq_gamma),
        ("equity", "vega", eqv_ws, eq_rho, eq_gamma),
        ("fx", "delta", fx_ws, fx_rho, fx_gamma),
    ]
    for rc, measure, ws, rho, gamma in dv_specs:
        for scen in SCENARIOS:
            res = delta_vega_charge(ws, rho, gamma, scen, hi, lo)
            charges[rc][measure][scen] = res.charge
            if scen == "medium":
                kb_medium[rc][measure] = dict(res.kb)

    # -- curvature ----------------------------------------------------------
    def one_factor_keys(cvr: Mapping[str, Tuple[float, float]]) -> Dict[str, List[str]]:
        return {b: ["crv"] for b in cvr}

    girr_cvr = {ccy: ([up], [dn]) for ccy, (up, dn) in sens.girr_cvr.items()}
    eq_cvr: Dict[str, Tuple[List[float], List[float]]] = {}
    eq_keys: Dict[str, List[str]] = {}
    for name, (up, dn) in sens.equity_cvr.items():
        b = market.equity(name).bucket
        params.equity_bucket(b)
        if b not in eq_cvr:
            eq_cvr[b] = ([], [])
            eq_keys[b] = []
        eq_cvr[b][0].append(up)
        eq_cvr[b][1].append(dn)
        eq_keys[b].append(name)
    fx_cvr = {"FX": ([u for u, _ in sens.fx_cvr.values()],
                     [d for _, d in sens.fx_cvr.values()])} if sens.fx_cvr else {}
    fx_keys = {"FX": sorted(sens.fx_cvr)}

    crv_specs = [
        ("girr", girr_cvr, girr_rho, girr_gamma, one_factor_keys(sens.girr_cvr)),
        ("equity", eq_cvr, eq_rho, eq_gamma, eq_keys),
        ("fx", fx_cvr, fx_rho, fx_gamma, fx_keys),
    ]
    for rc, cvr, rho, gamma, keys in crv_specs:
        for scen in SCENARIOS:
            res = curvature_charge(cvr, rho, gamma, scen, keys, hi, lo)
            charges[rc]["curvature"][scen] = res.charge
            if scen == "medium":
                kb_medium[rc]["curvature"] = dict(res.kb)

    # fx vega not modelled: pin to zero for all scenarios
    for scen in SCENARIOS:
        charges["fx"]["vega"][scen] = 0.0

    scenario_totals = {
        scen: sum(charges[rc][m].get(scen, 0.0) for rc in RISK_CLASSES for m in MEASURES)
        for scen in SCENARIOS
    }
    capital = max(scenario_totals.values())
    return SbmResult(charges=charges, kb_medium=kb_medium,
                     scenario_totals=scenario_totals, capital=capital)


# --------------------------------------------------------------------------
# DRC-lite
# --------------------------------------------------------------------------

@dataclass(frozen=True)
class DrcPosition:
    """One default-risk position: issuer, rating, notional (signed), market value."""

    issuer: str
    rating: str
    notional: float
    market_value: float
    lgd: float = 0.75

    def jtd(self) -> float:
        """Jump-to-default: JTD = LGD*notional + (MV - notional) (signed)."""
        return self.lgd * self.notional + (self.market_value - self.notional)


@dataclass(frozen=True)
class DrcResult:
    """DRC-lite output: charge + netting/HBR drill-down."""

    charge: float
    hbr: float
    net_jtd: Dict[str, float]
    gross_long: float
    gross_short: float


def drc_charge(positions: Sequence[DrcPosition], params: SbmParams) -> DrcResult:
    """Default Risk Charge (lite).

    1. JTD_i = LGD*notional + (MV - notional) per position (signed; shorts have
       negative notional).
    2. Net JTD per issuer (long/short netting within the same issuer).
    3. HBR = sum(netLong) / (sum(netLong) + sum(|netShort|)); HBR = 1 when
       there are no net shorts (all-long edge case) and when the book is empty.
    4. DRC = max(0, sum RW_i*netLong_i - HBR * sum RW_i*|netShort_i|)
       with RW from the pinned rating table (unknown rating -> ValueError).
    """
    net: Dict[str, float] = {}
    rating_of: Dict[str, str] = {}
    for p in positions:
        net[p.issuer] = net.get(p.issuer, 0.0) + p.jtd()
        prev = rating_of.setdefault(p.issuer, p.rating)
        if prev != p.rating:
            raise ValueError(f"drc_charge: issuer '{p.issuer}' has inconsistent ratings "
                             f"('{prev}' vs '{p.rating}')")
    long_sum = sum(v for v in net.values() if v > 0.0)
    short_sum = sum(-v for v in net.values() if v < 0.0)
    denom = long_sum + short_sum
    hbr = long_sum / denom if denom > 0.0 else 1.0
    weighted_long = sum(params.drc_rw(rating_of[i]) * v for i, v in net.items() if v > 0.0)
    weighted_short = sum(params.drc_rw(rating_of[i]) * (-v) for i, v in net.items() if v < 0.0)
    charge = max(0.0, weighted_long - hbr * weighted_short)
    return DrcResult(charge=charge, hbr=hbr, net_jtd=net,
                     gross_long=long_sum, gross_short=short_sum)


def drc_positions_from_instruments(instruments: Sequence[Instrument],
                                   market: Market) -> List[DrcPosition]:
    """Extract DRC positions (bonds only in this educational kit — documented)."""
    out: List[DrcPosition] = []
    for inst in instruments:
        if isinstance(inst, Bond):
            mv = price_bond(inst, market.curve(inst.currency))
            out.append(DrcPosition(issuer=inst.issuer, rating=inst.rating,
                                   notional=inst.notional, market_value=mv, lgd=inst.lgd))
    return out


# --------------------------------------------------------------------------
# RRAO
# --------------------------------------------------------------------------

def rrao_charge(instruments: Sequence[Instrument], params: SbmParams) -> float:
    """Residual risk add-on: sum over flagged instruments of rate * notional.

    Pinned rates: 'exotic' 1.0%, 'other' 0.1% of the flagged notional.
    """
    total = 0.0
    for inst in instruments:
        flag = getattr(inst, "rrao", None)
        if flag is not None:
            total += params.rrao_rate(flag.category) * flag.notional
    return total

File: java/src/main/java/com/quant/frtb/Sa.java
package com.quant.frtb;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Standardised Approach assembly: SBM charges per risk class and scenario,
 * plus DRC-lite and RRAO. SA capital = SBM + DRC + RRAO.
 */
public final class Sa {

    private Sa() {
    }

    /** Risk classes in reporting order. */
    public static final List<String> RISK_CLASSES = List.of("girr", "equity", "fx");

    /** Measures in reporting order. */
    public static final List<String> MEASURES = List.of("delta", "vega", "curvature");

    /**
     * SBM capital with full drill-down.
     *
     * @param charges        {@code riskClass -> measure -> scenario -> charge}
     * @param kbMedium       {@code riskClass -> measure -> bucket -> K_b} (medium scenario)
     * @param scenarioTotals {@code scenario -> sum over risk classes and measures}
     * @param capital        max over scenarios of the scenario totals
     */
    public record SbmResult(Map<String, Map<String, Map<String, Double>>> charges,
                            Map<String, Map<String, Map<String, Double>>> kbMedium,
                            Map<String, Double> scenarioTotals,
                            double capital) {
    }

    /** Format a tenor as its Python {@code "%g"} label ("0.25", "1", "10", ...). */
    static String tenorLabel(double t) {
        if (t == Math.rint(t) && Math.abs(t) < 1e15) {
            return Long.toString((long) t);
        }
        return Double.toString(t);
    }

    /** Assemble the full SBM capital: 3 risk classes x 3 measures x 3 scenarios. */
    public static SbmResult sbmCapital(Sensitivities sens, Market market, SbmParams params) {
        double hi = params.scenarioHigh();
        double lo = params.scenarioLow();
        Map<String, Map<String, Map<String, Double>>> charges = new LinkedHashMap<>();
        Map<String, Map<String, Map<String, Double>>> kbMedium = new LinkedHashMap<>();
        for (String rc : RISK_CLASSES) {
            Map<String, Map<String, Double>> perMeasure = new LinkedHashMap<>();
            for (String m : MEASURES) {
                perMeasure.put(m, new LinkedHashMap<>());
            }
            charges.put(rc, perMeasure);
            kbMedium.put(rc, new LinkedHashMap<>());
        }

        // -- GIRR delta structs ------------------------------------------------
        double[] tenors = params.girrTenors();
        Map<String, Integer> labelIndex = new LinkedHashMap<>();
        for (int i = 0; i < tenors.length; i++) {
            labelIndex.put(tenorLabel(tenors[i]), i);
        }
        Map<String, Map<String, Double>> girrWs = new LinkedHashMap<>();
        for (Map.Entry<String, Map<Double, Double>> e : sens.girr().entrySet()) {
            Map<String, Double> perTenor = new LinkedHashMap<>();
            for (Map.Entry<Double, Double> te : e.getValue().entrySet()) {
                perTenor.put(tenorLabel(te.getKey()),
                        params.girrRw(te.getKey()) * te.getValue());
            }
            girrWs.put(e.getKey(), perTenor);
        }
        Sbm.IntraRho girrRho = (bucket, k, l) ->
                params.girrRhoKl(labelIndex.get(k), labelIndex.get(l));
        Sbm.CrossGamma girrGamma = (b, c) -> params.girrGamma();

        // -- Equity delta/vega structs ----------------------------------------
        Map<String, Map<String, Double>> eqdWs = equityWs(sens.equityDelta(), market, params, false);
        Map<String, Map<String, Double>> eqvWs = equityWs(sens.equityVega(), market, params, true);
        Sbm.IntraRho eqRho = (bucket, k, l) -> params.equityBucket(bucket).rho();
        Sbm.CrossGamma eqGamma = (b, c) -> params.equityGamma();

        // -- FX delta structs (single pinned "FX" bucket) ---------------------
        Map<String, Map<String, Double>> fxWs = new LinkedHashMap<>();
        if (!sens.fxDelta().isEmpty()) {
            Map<String, Double> perPair = new LinkedHashMap<>();
            for (Map.Entry<String, Double> e : sens.fxDelta().entrySet()) {
                perPair.put(e.getKey(), params.fxDeltaRw() * e.getValue());
            }
            fxWs.put("FX", perPair);
        }
        Sbm.IntraRho fxRho = (bucket, k, l) -> params.fxRho();
        Sbm.CrossGamma fxGamma = (b, c) -> params.fxGamma();

        // -- delta / vega ------------------------------------------------------
        record DvSpec(String rc, String measure, Map<String, Map<String, Double>> ws,
                      Sbm.IntraRho rho, Sbm.CrossGamma gamma) {
        }
        List<DvSpec> dvSpecs = List.of(
                new DvSpec("girr", "delta", girrWs, girrRho, girrGamma),
                // no IR-vol instruments in scope => GIRR vega is identically 0
                new DvSpec("girr", "vega", Map.of(), girrRho, girrGamma),
                new DvSpec("equity", "delta", eqdWs, eqRho, eqGamma),
                new DvSpec("equity", "vega", eqvWs, eqRho, eqGamma),
                new DvSpec("fx", "delta", fxWs, fxRho, fxGamma));
        for (DvSpec spec : dvSpecs) {
            for (String scen : Sbm.SCENARIOS) {
                Sbm.RiskClassCharge res = Sbm.deltaVegaCharge(spec.ws(), spec.rho(),
                        spec.gamma(), scen, hi, lo);
                charges.get(spec.rc()).get(spec.measure()).put(scen, res.charge());
                if ("medium".equals(scen)) {
                    kbMedium.get(spec.rc()).put(spec.measure(), new LinkedHashMap<>(res.kb()));
                }
            }
        }

        // -- curvature ---------------------------------------------------------
        Map<String, List<List<Double>>> girrCvr = new LinkedHashMap<>();
        Map<String, List<String>> girrKeys = new LinkedHashMap<>();
        for (Map.Entry<String, double[]> e : sens.girrCvr().entrySet()) {
            girrCvr.put(e.getKey(), List.of(List.of(e.getValue()[0]), List.of(e.getValue()[1])));
            girrKeys.put(e.getKey(), List.of("crv"));
        }
        Map<String, List<List<Double>>> eqCvr = new LinkedHashMap<>();
        Map<String, List<String>> eqKeys = new LinkedHashMap<>();
        for (Map.Entry<String, double[]> e : sens.equityCvr().entrySet()) {
            String b = market.equity(e.getKey()).bucket();
            params.equityBucket(b); // error when the bucket is not pinned
            List<List<Double>> lists = eqCvr.computeIfAbsent(b,
                    key -> List.of(new ArrayList<>(), new ArrayList<>()));
            eqKeys.computeIfAbsent(b, key -> new ArrayList<>()).add(e.getKey());
            lists.get(0).add(e.getValue()[0]);
            lists.get(1).add(e.getValue()[1]);
        }
        Map<String, List<List<Double>>> fxCvr = new LinkedHashMap<>();
        Map<String, List<String>> fxKeys = new LinkedHashMap<>();
        if (!sens.fxCvr().isEmpty()) {
            List<Double> ups = new ArrayList<>();
            List<Double> dns = new ArrayList<>();
            List<String> keys = new ArrayList<>();
            for (Map.Entry<String, double[]> e : sens.fxCvr().entrySet()) {
                ups.add(e.getValue()[0]);
                dns.add(e.getValue()[1]);
                keys.add(e.getKey());
            }
            fxCvr.put("FX", List.of(ups, dns));
            fxKeys.put("FX", keys);
        }

        record CrvSpec(String rc, Map<String, List<List<Double>>> cvr, Sbm.IntraRho rho,
                       Sbm.CrossGamma gamma, Map<String, List<String>> keys) {
        }
        List<CrvSpec> crvSpecs = List.of(
                new CrvSpec("girr", girrCvr, girrRho, girrGamma, girrKeys),
                new CrvSpec("equity", eqCvr, eqRho, eqGamma, eqKeys),
                new CrvSpec("fx", fxCvr, fxRho, fxGamma, fxKeys));
        for (CrvSpec spec : crvSpecs) {
            for (String scen : Sbm.SCENARIOS) {
                Sbm.RiskClassCharge res = Sbm.curvatureCharge(spec.cvr(), spec.rho(),
                        spec.gamma(), scen, spec.keys(), hi, lo);
                charges.get(spec.rc()).get("curvature").put(scen, res.charge());
                if ("medium".equals(scen)) {
                    kbMedium.get(spec.rc()).put("curvature", new LinkedHashMap<>(res.kb()));
                }
            }
        }

        // fx vega not modelled: pin to zero for all scenarios
        for (String scen : Sbm.SCENARIOS) {
            charges.get("fx").get("vega").put(scen, 0.0);
        }

        Map<String, Double> scenarioTotals = new LinkedHashMap<>();
        for (String scen : Sbm.SCENARIOS) {
            double total = 0.0;
            for (String rc : RISK_CLASSES) {
                for (String m : MEASURES) {
                    Double c = charges.get(rc).get(m).get(scen);
                    total += c == null ? 0.0 : c;
                }
            }
            scenarioTotals.put(scen, total);
        }
        double capital = scenarioTotals.values().stream()
                .mapToDouble(Double::doubleValue).max().orElse(0.0);
        return new SbmResult(charges, kbMedium, scenarioTotals, capital);
    }

    /** Equity delta or vega WS per bucket, factor key = underlier name. */
    private static Map<String, Map<String, Double>> equityWs(Map<String, Double> sensMap,
                                                             Market market, SbmParams params,
                                                             boolean vega) {
        Map<String, Map<String, Double>> bucketWs = new LinkedHashMap<>();
        for (Map.Entry<String, Double> e : sensMap.entrySet()) {
            String b = market.equity(e.getKey()).bucket();
            SbmParams.EquityBucket p = params.equityBucket(b); // error when not pinned
            double rw = vega ? p.vegaRw() : p.deltaRw();
            bucketWs.computeIfAbsent(b, key -> new LinkedHashMap<>())
                    .put(e.getKey(), rw * e.getValue());
        }
        return bucketWs;
    }

    // ---------------------------------------------------------------------
    // DRC-lite
    // ---------------------------------------------------------------------

    /**
     * One default-risk position: issuer, rating, notional (signed), market
     * value; {@code lgd} pinned at 0.75 for bonds.
     *
     * @param issuer      netting key
     * @param rating      pinned rating bucket
     * @param notional    signed face amount (shorts negative)
     * @param marketValue current mark-to-market
     * @param lgd         loss-given-default
     */
    public record DrcPosition(String issuer, String rating, double notional,
                              double marketValue, double lgd) {

        /** Jump-to-default: {@code JTD = LGD*notional + (MV - notional)} (signed). */
        public double jtd() {
            return lgd * notional + (marketValue - notional);
        }
    }

    /**
     * DRC-lite output: charge plus netting/HBR drill-down.
     *
     * @param charge     the default risk charge
     * @param hbr        hedge benefit ratio (1 when no net shorts)
     * @param netJtd     net JTD per issuer
     * @param grossLong  sum of net long JTDs
     * @param grossShort sum of |net short JTDs|
     */
    public record DrcResult(double charge, double hbr, Map<String, Double> netJtd,
                            double grossLong, double grossShort) {
    }

    /**
     * Default Risk Charge (lite).
     *
     * <ol>
     *   <li>{@code JTD_i = LGD*notional + (MV - notional)} per position.</li>
     *   <li>Net JTD per issuer (long/short netting within the same issuer).</li>
     *   <li>{@code HBR = sum(netLong) / (sum(netLong) + sum(|netShort|))};
     *       HBR = 1 with no net shorts and for an empty book.</li>
     *   <li>{@code DRC = max(0, sum RW*netLong - HBR * sum RW*|netShort|)}
     *       with RW from the pinned rating table (unknown rating raises).</li>
     * </ol>
     */
    public static DrcResult drcCharge(List<DrcPosition> positions, SbmParams params) {
        Map<String, Double> net = new LinkedHashMap<>();
        Map<String, String> ratingOf = new LinkedHashMap<>();
        for (DrcPosition p : positions) {
            net.merge(p.issuer(), p.jtd(), Double::sum);
            String prev = ratingOf.putIfAbsent(p.issuer(), p.rating());
            if (prev != null && !prev.equals(p.rating())) {
                throw new IllegalArgumentException("drcCharge: issuer '" + p.issuer()
                        + "' has inconsistent ratings ('" + prev + "' vs '" + p.rating() + "')");
            }
        }
        double longSum = 0.0;
        double shortSum = 0.0;
        for (double v : net.values()) {
            if (v > 0.0) {
                longSum += v;
            } else if (v < 0.0) {
                shortSum += -v;
            }
        }
        double denom = longSum + shortSum;
        double hbr = denom > 0.0 ? longSum / denom : 1.0;
        double weightedLong = 0.0;
        double weightedShort = 0.0;
        for (Map.Entry<String, Double> e : net.entrySet()) {
            if (e.getValue() > 0.0) {
                weightedLong += params.drcRw(ratingOf.get(e.getKey())) * e.getValue();
            } else if (e.getValue() < 0.0) {
                weightedShort += params.drcRw(ratingOf.get(e.getKey())) * (-e.getValue());
            }
        }
        double charge = Math.max(0.0, weightedLong - hbr * weightedShort);
        return new DrcResult(charge, hbr, net, longSum, shortSum);
    }

    /** Extract DRC positions (bonds only in this educational kit — documented). */
    public static List<DrcPosition> drcPositionsFromInstruments(List<Instrument> instruments,
                                                                Market market) {
        List<DrcPosition> out = new ArrayList<>();
        for (Instrument inst : instruments) {
            if (inst instanceof Bond bond) {
                double mv = Pricers.priceBond(bond, market.curve(bond.currency()));
                out.add(new DrcPosition(bond.issuer(), bond.rating(), bond.notional(),
                        mv, bond.lgd()));
            }
        }
        return out;
    }

    // ---------------------------------------------------------------------
    // RRAO
    // ---------------------------------------------------------------------

    /**
     * Residual risk add-on: sum over flagged instruments of rate * notional.
     * Pinned rates: 'exotic' 1.0%, 'other' 0.1% of the flagged notional.
     */
    public static double rraoCharge(List<Instrument> instruments, SbmParams params) {
        double total = 0.0;
        for (Instrument inst : instruments) {
            RraoFlag flag = inst.rrao();
            if (flag != null) {
                total += params.rraoRate(flag.category()) * flag.notional();
            }
        }
        return total;
    }
}
